packages/app/src/cli/services/function/ui/components/hooks/extension-watcher.ts

```diff
@@ -0,0 +1,193 @@
+
+interface WatchFunctionForReplayOptions {
+  selectedRun: FunctionRunData
+  abortController: AbortController
+  app: AppInterface
+  extension: ExtensionInstance<FunctionConfigType>
+}
+
+export function setupExtensionWatcherForReplay({
+  selectedRun,
+  abortController,
+  app,
+  extension,
+}: WatchFunctionForReplayOptions) {
+  const functionRunFromSelectedRun = {
+    type: 'functionRun',
+    input: selectedRun.payload.input,
+    output: selectedRun.payload.output,
+    logs: selectedRun.payload.logs,
+    name: selectedRun.source,
+    size: 0,
+    memory_usage: 0,
+    instructions: selectedRun.payload.fuelConsumed,
+  } as FunctionRun
+
+  const [logs, setLogs] = useState<ReplayLog[]>([])
+  const [recentFunctionRuns, setRecentFunctionRuns] = useState<[FunctionRun, FunctionRun]>([
+    functionRunFromSelectedRun,
+    functionRunFromSelectedRun,
+  ])
+
+  const [error, setError] = useState<string | undefined>(undefined)
+
+  const {input, export: runExport} = selectedRun.payload
+
+  const {isRawModeSupported: canUseShortcuts} = useStdin()
+  const pollingInterval = useRef<NodeJS.Timeout>()
+
+  const [statusMessage, setStatusMessage] = useState(`Watching for changes to ${selectedRun.source}...`)
+
+  useEffect(() => {
+    // run the selectedRun once
+    const initialReplay = async () => {
+      setStatusMessage('Replaying log with local function...')
+      const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
+      setRecentFunctionRuns((recentFunctionRuns) => {
+        return [functionRun, recentFunctionRuns[0]]
+      })
+      setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
+      setLogs((logs) => [...logs, functionRun])
+    }
+
+    const startWatchingFunction = async () => {
+      const customStdout = new Writable({
+        write(chunk, _enconding, next) {
+          setLogs((logs) => [...logs, {type: 'systemMessage', message: chunk.toString()}])
+          next()
+        },
+      })
+
+      await setupExtensionWatcher({
+        extension,
+        app,
+        stdout: customStdout,
+        stderr: customStdout,
+        onChange: async () => {
+          setError('')
+          setStatusMessage('Re-running with latest changes...')
+          const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
+
+          setRecentFunctionRuns((recentFunctionRuns) => {
+            return [functionRun, recentFunctionRuns[0]]
+          })
+
+          setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
+          setLogs((logs) => [...logs, functionRun])
+        },
+        onReloadAndBuildError: async (error) => {
+          setError('Error while reloading and building extension')
+          if (error instanceof FatalError) {
+            renderFatalError(error)
+          } else {
+            outputWarn(`Failed to replay function: ${error.message}`)
```

Comment:
If we changed this to `setStatusMessage` instead of `outputWarn`, would we still need the `customStdout`?